CI [Testing]

cjyothika opened the pull request:
Define Github Actions workflows for implenting CI for the repository.

Added testing for frontend/backend, using the Jasmine framework.

Test cases should follow the following format as specified by the [Jasmine documentation](https://jasmine.github.io/):

```
describe("A suite is just a function", function() {
  let a;

  it("and so is a spec", function() {
    a = true;

    expect(a).toBe(true);
  });
});
```

cjyothika committed:
Create docker-image.yml

cjyothika committed:
Create docker-image.yml

chia-yh commented:
The check seems to be failing? Is it possible to have it run docker-compose with the file in the root and see if that succeeds? Or should we do both compose and build for each container

Commit:
trying to fix CI

Commit:
trying to fix CI

Commit:
adding frontend tests

Commit:
adding frontend tests

Commit:
trying to fix frontend ci

Commit:
trying to fix frontend ci

Commit:
fixing more issues

Commit:
fixing more issues

Commit:
trying something

Commit:
trying something

Commit:
trying

Commit:
trying

Commit:
pray

Commit:
pray

Commit:
collab service still fails

Commit:
collab service still fails

GenFusion122 committed:
Implement unit tests

Implemented unit tests on each backend service using Jasmine.

GenFusion122 committed:
Implement more modular unit tests and minor improvements

All tests now use `app.spec.js` to run the server as opposed to running their respective index.js.

Added a timeout of 1 second to unit tests for matching service, this should fix the issue of it failing occasionally.

Improved some tests for question service.

Commit:
fixing a test

Commit:
fixing a test

Commit:
pray this works

Commit:
pray this works

Commit:
trying something

Commit:
trying something

Commit:
name change

Commit:
name change

GenFusion122 commented on .github/workflows/main.yml:
```diff
@@ -0,0 +1,32 @@
+      run: |
+          npm install
+          npm install -g @angular/cli
+          xvfb-run -a ng test --watch=false --browsers=ChromeHeadless 
```
Can I check if these are the commands you'd use to run the tests? 
Does it work for Jasmine tests in the backend too?

itsrx replied:
Yea these were the commands to run the tests. I'm pretty sure you just paste the code you use to run the backend tests and it should work. The only reason why I had to do `xvfb-run -a ng test --watch=false --browsers=ChromeHeadless` was so that the tests can run on a "fake" chrome browser

chia-yh commented on frontend/src/app/app.component.spec.ts:
```diff
@@ -18,10 +21,10 @@ describe('AppComponent', () => {
-    const compiled = fixture.nativeElement as HTMLElement;
-    expect(compiled.querySelector('.content span')?.textContent).toContain('frontend app is running!');
-  });
+  // it('should render title', () => {
```
Would like to confirm if this should be left in here @itsrx @GenFusion122

itsrx replied:
Oops, I will delete it. The app doesn't say this anymore so this test is redundant

Commit:
removing comment

Commit:
removing comment

GenFusion122 reviewed (commented):
I think overall it LGTM, but we might wanna consider pushing to a different branch with more updated code or update this branch first? I know the code base on this branch is pretty old.

Commit:
fixing test cases based on ui updates

itsrx committed:
Merge branch 'ci-cd' into ci-cd-backup

itsrx committed:
Merge pull request #53 from CS3219-AY2324S1/ci-cd-backup

Ci cd backup

Commit:
fixing testcase bugs

itsrx commented:
> I think overall it LGTM, but we might wanna consider pushing to a different branch with more updated code or update this branch first? I know the code base on this branch is pretty old.

pulled recent changes. Lmk if it's good to merge

itsrx commented:
> > I think overall it LGTM, but we might wanna consider pushing to a different branch with more updated code or update this branch first? I know the code base on this branch is pretty old.
> 
> pulled recent changes. Lmk if it's good to merge

the changes are based off of the ui-updates branch

chia-yh commented:
@GenFusion122 Do review, if its good to go we can merge it into main

GenFusion122 commented:
I'll go merge my unit tests into this branch too, so we can have a collective "cicd" branch. Once that's done and it works fine here then we can merge it into main together.

GenFusion122 committed:
Merge branch 'master' into testing

GenFusion122 committed:
Merge branch 'ci-cd' into testing

GenFusion122 committed:
Implement History unit tests + minor improvements + backend tests to CI

GenFusion122 committed:
Update CI

GenFusion122 committed:
Add more tests to main.yml

GenFusion122 committed:
Implement tags to question-service unit tests

GenFusion122 committed:
Merge pull request #2 from GenFusion122/testing

CI workflow testing

GenFusion122 committed:
Merge pull request #54 from GenFusion122/master

Implement unit tests for all services + backend

GenFusion122 reviewed (dismissed):
Lgtm

chia-yh commented on backend/matching-service/spec/matching-queue.spec.js:
```diff
@@ -0,0 +1,75 @@
+      expect(queue.getQueueLen()).toEqual(1);
+    });
+
+    it('returns undefined if the user already exists in the queue', () => {
```
Should we standardise the message here? As i recall, the jasmine standard was:

`describe({something})`

`it(`should {do something}')`

GenFusion122 replied:
Yup, changed!

chia-yh commented on backend/matching-service/spec/matching-queue.spec.js:
```diff
@@ -0,0 +1,75 @@
+    queue = new MatchingQueue();
+  });
+
+  describe('#enqueue', () => {
```
Should we also standardise between:

`describe('#{function_name}')`

and

`describe('{function_name} function')`

?

GenFusion122 replied:
Changed in the most recent commit!

chia-yh reviewed (changes requested):
Minor changes for consistency, after which it LGTM

GenFusion122 committed:
Update names of tests

GenFusion122 committed:
Merge branch 'testing' into ci-cd

GenFusion122 committed:
Merge branch 'master' into ci-cd

Commit:
Delete Collab Tests

cjyothika committed:
Merge branch 'master' into ci-cd

Commit:
Delete Dialog Tests

Commit:
Merge branch 'ci-cd' of https://github.com/CS3219-AY2324S1/ay2324s1-course-assessment-g40 into ci-cd
